// @netlify/content-engine/dist/structured-errors/error-schema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.errorSchema = exports.Position = void 0;
const joi_1 = __importDefault(require("joi"));
const types_1 = require("./types");
exports.Position = joi_1.default.object()
    .keys({
    line: joi_1.default.number(),
    column: joi_1.default.number(),
})
    .unknown();
exports.errorSchema = joi_1.default.object().keys({
    code: joi_1.default.string(),
    text: joi_1.default.string(),
    stack: joi_1.default.array()
        .items(joi_1.default.object().keys({
        fileName: joi_1.default.string(),
        functionName: joi_1.default.string().allow(null),
        lineNumber: joi_1.default.number().allow(null),
        columnNumber: joi_1.default.number().allow(null),
    }))
        .allow(null),
    category: joi_1.default.string().valid(...Object.values(types_1.ErrorCategory)),
    level: joi_1.default.string().valid(...Object.values(types_1.Level)),
    type: joi_1.default.string().valid(...Object.values(types_1.Type)),
    filePath: joi_1.default.string(),
    location: joi_1.default.object({
        start: exports.Position.required(),
        end: exports.Position,
    }).unknown(),
    docsUrl: joi_1.default.string().uri({
        allowRelative: false,
        relativeOnly: false,
    }),
    error: joi_1.default.object({}).unknown(),
    context: joi_1.default.object({}).unknown(),
    group: joi_1.default.string(),
    panicOnBuild: joi_1.default.boolean(),
    pluginName: joi_1.default.string(),
});
