// @netlify/content-engine/dist/structured-errors/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorCategory = exports.Type = exports.Level = void 0;
var Level;
(function (Level) {
    Level["ERROR"] = "ERROR";
    Level["WARNING"] = "WARNING";
    Level["INFO"] = "INFO";
    Level["DEBUG"] = "DEBUG";
})(Level || (exports.Level = Level = {}));
var Type;
(function (Type) {
    Type["HTML_COMPILATION"] = "HTML.COMPILATION";
    Type["HTML_GENERATION"] = "HTML.GENERATION";
    Type["HTML_GENERATION_DEV_SSR"] = "HTML.GENERATION.DEV_SSR";
    Type["HTML_GENERATION_SSG"] = "HTML.GENERATION.SSG";
    Type["RSC_COMPILATION"] = "RSC.COMPILATION";
    Type["RSC_UNKNOWN"] = "RSC.UNKNOWN";
    Type["PAGE_DATA"] = "PAGE_DATA";
    Type["GRAPHQL_SCHEMA"] = "GRAPHQL.SCHEMA";
    Type["GRAPHQL_QUERY_RUNNING"] = "GRAPHQL.QUERY_RUNNING";
    Type["GRAPHQL_EXTRACTION"] = "GRAPHQL.EXTRACTION";
    Type["GRAPHQL_VALIDATION"] = "GRAPHQL.VALIDATION";
    Type["GRAPHQL_UNKNOWN"] = "GRAPHQL.UNKNOWN";
    Type["ENGINE_COMPILATION"] = "ENGINE.COMPILATION";
    Type["ENGINE_HTML"] = "ENGINE.HTML";
    Type["ENGINE_VALIDATION"] = "ENGINE.VALIDATION";
    Type["ENGINE_EXECUTION"] = "ENGINE.EXECUTION";
    Type["API_CONFIG_VALIDATION"] = "API.CONFIG.VALIDATION";
    Type["API_CONFIG_LOADING"] = "API.CONFIG.LOADING";
    Type["API_CONFIG_COMPILATION"] = "API.CONFIG.COMPILATION";
    Type["API_NODE_VALIDATION"] = "API.NODE.VALIDATION";
    Type["API_NODE_COMPILATION"] = "API.NODE.COMPILATION";
    Type["API_NODE_EXECUTION"] = "API.NODE.EXECUTION";
    Type["API_TYPESCRIPT_COMPILATION"] = "API.TYPESCRIPT.COMPILATION";
    Type["API_TYPESCRIPT_TYPEGEN"] = "API.TYPESCRIPT.TYPEGEN";
    Type["FUNCTIONS_COMPILATION"] = "FUNCTIONS.COMPILATION";
    Type["FUNCTIONS_EXECUTION"] = "FUNCTIONS.EXECUTION";
    Type["CLI_VALIDATION"] = "CLI.VALIDATION";
    // webpack errors for each stage enum: packages/gatsby/src/commands/types.ts
    Type["WEBPACK_DEVELOP"] = "WEBPACK.DEVELOP";
    Type["WEBPACK_DEVELOP_HTML"] = "WEBPACK.DEVELOP-HTML";
    Type["WEBPACK_BUILD_JAVASCRIPT"] = "WEBPACK.BUILD-JAVASCRIPT";
    Type["WEBPACK_BUILD_HTML"] = "WEBPACK.BUILD-HTML";
    Type["UNKNOWN"] = "UNKNOWN";
    // Backwards compatibility for plugins
    // TODO(v6): Remove these types
    /** @deprecated */
    Type["GRAPHQL"] = "GRAPHQL";
    /** @deprecated */
    Type["CONFIG"] = "CONFIG";
    /** @deprecated */
    Type["WEBPACK"] = "WEBPACK";
    /** @deprecated */
    Type["PLUGIN"] = "PLUGIN";
    /** @deprecated */
    Type["COMPILATION"] = "COMPILATION";
})(Type || (exports.Type = Type = {}));
var ErrorCategory;
(function (ErrorCategory) {
    ErrorCategory["USER"] = "USER";
    ErrorCategory["SYSTEM"] = "SYSTEM";
    ErrorCategory["THIRD_PARTY"] = "THIRD_PARTY";
    ErrorCategory["UNKNOWN"] = "UNKNOWN";
})(ErrorCategory || (exports.ErrorCategory = ErrorCategory = {}));
