// @netlify/sdk/dist/cli/dev/open_extension_dev_url.js
import open from "open";
import { getToken } from "../../auth/get-token.js";
import { ntliLog } from "../../utils.js";
export async function openExtensionDevUrl(extensionDevUrl, slug, redirect) {
    const netlifyCliUserAuthToken = await getToken();
    if (!netlifyCliUserAuthToken?.[0]) {
        console.error("You are not logged in to Netlify. Please run `netlify login` to use `netlify-extension dev --open`.");
        return;
    }
    const appOrigin = getAppOrigin();
    const res = await fetch(`${appOrigin}/.netlify/functions/ntli-extension-dev-auth`, {
        method: "POST",
        headers: {
            authorization: `Bearer ${netlifyCliUserAuthToken?.[0]}`,
        },
        body: JSON.stringify({
            slug: slug,
            url: extensionDevUrl,
            redirect_path: redirect,
        }),
    });
    if (!res.ok) {
        const body = await res.text();
        console.error(`Failed to get dev url: ${res.status} ${body}`);
        return;
    }
    const body = (await res.json());
    if (!body.data?.token) {
        console.error("Failed to get dev url: missing token");
        return;
    }
    const url = new URL(`${appOrigin}/extension-ui/dev`);
    url.search = new URLSearchParams({
        "nf-extension-token": body.data.token,
    }).toString();
    ntliLog(`Opening Extension UI in the Netlify App at ${url.toString()}`);
    // open( .. { wait: true } causes the open subprocess to hang around forever, creating a memory leak of about 20MB per code save..
    // seems to be related to https://github.com/sindresorhus/open/issues/197
    open(url.toString());
}
function getAppOrigin() {
    const productionAppOrigin = "https://app.netlify.com";
    if (!process.env.UNSTABLE__DEV_NETLIFY_APP_URL?.startsWith?.("http")) {
        return productionAppOrigin;
    }
    try {
        return new URL(process.env.UNSTABLE__DEV_NETLIFY_APP_URL)?.origin;
    }
    catch (e) {
        console.error("error parsing UNSTABLE__DEV_NETLIFY_APP_URL", e);
    }
    return productionAppOrigin;
}
