// @netlify/sdk/dist/extension_configuration/schema.js
import { Surfaces } from "@netlify/sdk--ui-core";
import { z } from "zod";
export const ExtensionConfigurationSchema = z.object({
    name: z.string().optional(),
    description: z.string().optional(),
    slug: z.string().regex(/^[a-z0-9-]+$/, "slug must be lowercase with dashes"),
    hostSiteId: z.string().optional(),
    scopes: z
        .object({
        all: z.boolean().optional(),
        site: z.array(z.enum(["read", "write"])).optional(),
        env: z.array(z.enum(["read", "write", "delete"])).optional(),
        user: z.array(z.enum(["read", "write"])).optional(),
    })
        .optional(),
    use_cjs_shims: z.boolean().optional(),
    connector: z
        .object({
        "local-frontend": z
            .object({
            port: z.number(),
            command: z.string(),
            directory: z.string(),
        })
            .optional(),
    })
        .optional(),
    ui: z
        .object({
        build: z
            .object({ command: z.string().min(1).trim().optional() })
            .optional(),
        dev: z
            .object({ command: z.string().min(1).trim().optional() })
            .optional(),
        surfaces: z
            .array(z.nativeEnum(Surfaces))
            .refine((surfaces) => new Set(surfaces).values())
            .optional(),
    })
        .optional(),
});
export const ExtensionConfigurationFileSchema = z.object({
    config: ExtensionConfigurationSchema,
});
