// @netlify/sdk/dist/utils.js
import chalk from "chalk";
export function catchMaybePromise(maybePromise, onError, cleanup) {
    if (
    // much faster than instanceof
    maybePromise &&
        typeof maybePromise === `object` &&
        `catch` in maybePromise) {
        return maybePromise.catch(onError).finally(cleanup);
    }
    else {
        try {
            return maybePromise;
        }
        catch (e) {
            onError(e);
        }
        finally {
            cleanup();
        }
    }
}
export const ntliLog = (message, name = `Netlify SDK`) => {
    console.log(`${chalk.blue(name)} ${message}\n`);
};
export const printResetHeader = () => {
    // leave a marker line for last output
    console.log(chalk.bold.yellow(`─`.repeat(process.stdout.columns || 4)));
    // push all terminal output out of view
    console.log(`\n`.repeat(process.stdout.rows || 4));
};
export const envVarFromSlug = (slug) => {
    const safeSlug = slug.toUpperCase().replaceAll(/[^A-Z_]/g, "_");
    if (safeSlug.startsWith("_")) {
        return `I${safeSlug}`;
    }
    return safeSlug;
};
export const parseNonUndefined = (value) => {
    if (value === undefined) {
        throw new TypeError("Expected value to not be undefined but received: undefined");
    }
    return value;
};
const concatenateTemplateLiteralTag = (raw, ...keys) => (keys.length === 0 ? raw[0] : String.raw({ raw }, ...keys));
/**
 * Enables Prettier formatting on GQL blocks
 * https://prettier.io/docs/en/options.html#embedded-language-formatting
 */
export const gql = concatenateTemplateLiteralTag;
