// @netlify/content-engine/dist/redux/reducers/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.telemetry = exports.statefulSourcePlugins = exports.requestHeaders = exports.inferenceMetadata = exports.logs = exports.schemaCustomization = exports.jobsV2 = exports.jobs = exports.status = exports.schema = exports.config = exports.flattenedPlugins = exports.lastAction = exports.nodesTouched = exports.resolvedNodesCache = exports.nodesByType = exports.nodes = exports.typeOwners = exports.program = exports.definitions = void 0;
const type_owners_1 = require("./type-owners");
Object.defineProperty(exports, "typeOwners", { enumerable: true, get: function () { return type_owners_1.typeOwnersReducer; } });
const nodes_1 = require("./nodes");
Object.defineProperty(exports, "nodes", { enumerable: true, get: function () { return nodes_1.nodesReducer; } });
const logs_1 = require("../../reporter/redux/reducers/logs");
Object.defineProperty(exports, "logs", { enumerable: true, get: function () { return logs_1.reducer; } });
const schema_1 = require("./schema");
Object.defineProperty(exports, "schema", { enumerable: true, get: function () { return schema_1.schemaReducer; } });
const definitions_1 = require("./definitions");
Object.defineProperty(exports, "definitions", { enumerable: true, get: function () { return definitions_1.definitionsReducer; } });
const status_1 = require("./status");
Object.defineProperty(exports, "status", { enumerable: true, get: function () { return status_1.statusReducer; } });
const config_1 = require("./config");
Object.defineProperty(exports, "config", { enumerable: true, get: function () { return config_1.configReducer; } });
const last_action_1 = require("./last-action");
Object.defineProperty(exports, "lastAction", { enumerable: true, get: function () { return last_action_1.lastActionReducer; } });
const jobsv2_1 = require("./jobsv2");
Object.defineProperty(exports, "jobsV2", { enumerable: true, get: function () { return jobsv2_1.jobsV2Reducer; } });
const jobs_1 = require("./jobs");
Object.defineProperty(exports, "jobs", { enumerable: true, get: function () { return jobs_1.jobsReducer; } });
const nodes_by_type_1 = require("./nodes-by-type");
Object.defineProperty(exports, "nodesByType", { enumerable: true, get: function () { return nodes_by_type_1.nodesByTypeReducer; } });
const program_1 = require("./program");
Object.defineProperty(exports, "program", { enumerable: true, get: function () { return program_1.programReducer; } });
const resolved_nodes_1 = require("./resolved-nodes");
Object.defineProperty(exports, "resolvedNodesCache", { enumerable: true, get: function () { return resolved_nodes_1.resolvedNodesCacheReducer; } });
const nodes_touched_1 = require("./nodes-touched");
Object.defineProperty(exports, "nodesTouched", { enumerable: true, get: function () { return nodes_touched_1.nodesTouchedReducer; } });
const flattened_plugins_1 = require("./flattened-plugins");
Object.defineProperty(exports, "flattenedPlugins", { enumerable: true, get: function () { return flattened_plugins_1.flattenedPluginsReducer; } });
const schema_customization_1 = require("./schema-customization");
Object.defineProperty(exports, "schemaCustomization", { enumerable: true, get: function () { return schema_customization_1.schemaCustomizationReducer; } });
const inference_metadata_1 = require("./inference-metadata");
Object.defineProperty(exports, "inferenceMetadata", { enumerable: true, get: function () { return inference_metadata_1.inferenceMetadataReducer; } });
const telemetry_1 = require("./telemetry");
Object.defineProperty(exports, "telemetry", { enumerable: true, get: function () { return telemetry_1.telemetryReducer; } });
const set_request_headers_1 = require("./set-request-headers");
Object.defineProperty(exports, "requestHeaders", { enumerable: true, get: function () { return set_request_headers_1.setRequestHeadersReducer; } });
const stateful_source_plugins_1 = require("./stateful-source-plugins");
Object.defineProperty(exports, "statefulSourcePlugins", { enumerable: true, get: function () { return stateful_source_plugins_1.statefulSourcePluginsReducer; } });
