// @netlify/content-engine/dist/reporter/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActivityTypes = exports.ActivityStatuses = exports.ActivityLogLevels = exports.LogLevels = exports.Actions = void 0;
var Actions;
(function (Actions) {
    Actions["LogAction"] = "LOG_ACTION";
    Actions["SetStatus"] = "SET_STATUS";
    Actions["Log"] = "LOG";
    Actions["SetLogs"] = "SET_LOGS";
    Actions["StartActivity"] = "ACTIVITY_START";
    Actions["EndActivity"] = "ACTIVITY_END";
    Actions["UpdateActivity"] = "ACTIVITY_UPDATE";
    Actions["PendingActivity"] = "ACTIVITY_PENDING";
    Actions["CancelActivity"] = "ACTIVITY_CANCEL";
    Actions["ActivityErrored"] = "ACTIVITY_ERRORED";
})(Actions || (exports.Actions = Actions = {}));
var LogLevels;
(function (LogLevels) {
    LogLevels["Debug"] = "DEBUG";
    LogLevels["Success"] = "SUCCESS";
    LogLevels["Info"] = "INFO";
    LogLevels["Warning"] = "WARNING";
    LogLevels["Log"] = "LOG";
    LogLevels["Error"] = "ERROR";
})(LogLevels || (exports.LogLevels = LogLevels = {}));
var ActivityLogLevels;
(function (ActivityLogLevels) {
    ActivityLogLevels["Success"] = "ACTIVITY_SUCCESS";
    ActivityLogLevels["Failed"] = "ACTIVITY_FAILED";
    ActivityLogLevels["Interrupted"] = "ACTIVITY_INTERRUPTED";
})(ActivityLogLevels || (exports.ActivityLogLevels = ActivityLogLevels = {}));
var ActivityStatuses;
(function (ActivityStatuses) {
    ActivityStatuses["InProgress"] = "IN_PROGRESS";
    ActivityStatuses["NotStarted"] = "NOT_STARTED";
    ActivityStatuses["Interrupted"] = "INTERRUPTED";
    ActivityStatuses["Failed"] = "FAILED";
    ActivityStatuses["Success"] = "SUCCESS";
    ActivityStatuses["Cancelled"] = "CANCELLED";
})(ActivityStatuses || (exports.ActivityStatuses = ActivityStatuses = {}));
var ActivityTypes;
(function (ActivityTypes) {
    ActivityTypes["Spinner"] = "spinner";
    ActivityTypes["Hidden"] = "hidden";
    ActivityTypes["Progress"] = "progress";
    ActivityTypes["Pending"] = "pending";
})(ActivityTypes || (exports.ActivityTypes = ActivityTypes = {}));
