// @netlify/content-engine/dist/reporter/reporter-timer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTimerReporter = void 0;
const constants_1 = require("./constants");
const createTimerReporter = ({ text, id, span, reporter, reporterActions, pluginName, }) => {
    return {
        start() {
            reporterActions.startActivity({
                id,
                text: text || `__timer__`,
                type: constants_1.ActivityTypes.Spinner,
            });
        },
        setStatus(statusText) {
            reporterActions.setActivityStatusText({
                id,
                statusText,
            });
        },
        panicOnBuild(errorMeta, error) {
            span.finish();
            reporterActions.setActivityErrored({
                id,
            });
            return reporter.panicOnBuild(errorMeta, error, pluginName);
        },
        panic(errorMeta, error) {
            span.finish();
            reporterActions.endActivity({
                id,
                status: constants_1.ActivityStatuses.Failed,
            });
            return reporter.panic(errorMeta, error, pluginName);
        },
        end() {
            span.finish();
            reporterActions.endActivity({
                id,
                status: constants_1.ActivityStatuses.Success,
            });
        },
        span,
    };
};
exports.createTimerReporter = createTimerReporter;
