// @netlify/sdk/dist/extension/connect/model/transformers/scalar/boolean.js
import { ModelError } from "../../error.js";
import { omitByUndefined } from "../../../utils/omit-by-undefined.js";
import { ScalarBaseTransformer } from "../scalar.js";
export class BooleanTransformer extends ScalarBaseTransformer {
    toCreateSingleFieldDefinition(modelField) {
        return omitByUndefined({
            ...this.modelFieldToSharedCreateProperties(modelField),
            type: this.visualEditorFieldType,
        });
    }
    toCreateListFieldDefinition(modelField) {
        return omitByUndefined({
            ...this.modelFieldToSharedCreateProperties(modelField),
            type: "list",
            items: {
                type: this.visualEditorFieldType,
            },
        });
    }
    parseVisitedInputSingle(input, field, treeContext) {
        const isBoolean = typeof input === `boolean`;
        if (!input && !field.required && !isBoolean) {
            return null;
        }
        if (!isBoolean) {
            throw new ModelError(`Input data for field ${field.name} at input path ${treeContext.path.join(`.`)}.${field.name} is not a boolean but should be. Received ${typeof input}.\n\n${JSON.stringify(input, null, 2)}`);
        }
        return input;
    }
    get visualEditorFieldType() {
        return "boolean";
    }
}
