// @netlify/sdk/dist/extension/connect/model/transformers/scalar/integer.js
import { ModelError } from "../../error.js";
import { omitByUndefined } from "../../../utils/omit-by-undefined.js";
import { ScalarBaseTransformer } from "../scalar.js";
export class IntegerTranformer extends ScalarBaseTransformer {
    parseNumberControlType(modelField) {
        const controlType = modelField.editor?.controlType;
        if (controlType && controlType !== `slider`) {
            const typeName = this.definition.getDefinedTypeName(modelField.type);
            throw new ModelError(`Control type ${controlType} is not a valid control type for field ${modelField.name} of type ${typeName}`);
        }
        return controlType;
    }
    toCreateSingleFieldDefinition(modelField) {
        return omitByUndefined({
            ...this.modelFieldToSharedCreateProperties(modelField),
            ...this.modelFieldToSharedCreateNumberProperties(modelField),
            controlType: this.parseNumberControlType(modelField),
            type: this.visualEditorFieldType,
            subtype: "int",
        });
    }
    toCreateListFieldDefinition(modelField) {
        return omitByUndefined({
            ...this.modelFieldToSharedCreateProperties(modelField),
            type: "list",
            items: {
                ...this.modelFieldToSharedCreateNumberProperties(modelField),
                controlType: this.parseNumberControlType(modelField),
                type: this.visualEditorFieldType,
                subtype: "int",
            },
        });
    }
    parseVisitedInputSingle(input, field, treeContext) {
        const isNumber = typeof input === `number`;
        if (!input && !field.required && !isNumber) {
            return null;
        }
        if (!isNumber) {
            throw new ModelError(`Input data for field ${field.name} at input path ${treeContext.path.join(`.`)}.${field.name} is not an integer but should be. Received ${typeof input}.\n\n${JSON.stringify(input, null, 2)}`);
        }
        return input;
    }
    get visualEditorFieldType() {
        return "number";
    }
}
