// @netlify/content-engine/dist/redux/persist.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.readFromCache = readFromCache;
exports.guessSafeChunkSize = guessSafeChunkSize;
exports.writeToCache = writeToCache;
const path_1 = __importDefault(require("path"));
const os_1 = __importDefault(require("os"));
const v8_1 = __importDefault(require("v8"));
const flatted_1 = require("flatted");
const fs_extra_1 = require("fs-extra");
const glob_1 = require("glob");
const reporter_1 = __importDefault(require("../reporter"));
const schema_customization_1 = require("./reducers/schema-customization");
const getReduxCacheFolder = () => 
// This is a function for the case that somebody does a process.chdir (#19800)
path_1.default.join(process.cwd(), `.cache/redux`);
function reduxSharedFile(dir) {
    return path_1.default.join(dir, `redux.rest.state`);
}
function reduxChunkedNodesFilePrefix(dir) {
    return path_1.default.join(dir, `redux.node.state_`);
}
function readFromCache() {
    // The cache is stored in two steps; the nodes and pages in chunks and the rest
    // First we revive the rest, then we inject the nodes and pages into that obj (if any)
    // Each chunk is stored in its own file, this circumvents max buffer lengths
    // for sites with a _lot_ of content. Since all nodes / pages go into a Map, the order
    // of reading them is not relevant.
    const cacheFolder = getReduxCacheFolder();
    const obj = v8_1.default.deserialize((0, fs_extra_1.readFileSync)(reduxSharedFile(cacheFolder)));
    // Note: at 1M pages, this will be 1M/chunkSize chunks (ie. 1m/10k=100)
    const nodesChunks = (0, glob_1.sync)(reduxChunkedNodesFilePrefix(cacheFolder) + `*`).map((file) => v8_1.default.deserialize((0, fs_extra_1.readFileSync)(file)));
    const nodes = [].concat(...nodesChunks);
    if (!nodesChunks.length) {
        reporter_1.default.verbose(`Cache exists but contains no nodes. There should be at least some nodes available so it seems the cache was corrupted. Disregarding the cache and proceeding as if there was none.`);
        return {};
    }
    obj.nodes = new Map(nodes);
    obj.schemaCustomization ||= {};
    const parsedSchemaCustomization = typeof obj.schemaCustomization === `string`
        ? (0, flatted_1.parse)(obj.schemaCustomization)
        : obj.schemaCustomization;
    obj.schemaCustomization = {
        ...(0, schema_customization_1.initialState)(),
        ...parsedSchemaCustomization,
    };
    return obj;
}
function guessSafeChunkSize(values, showMaxSizeWarning = false) {
    // Pick a few random elements and measure their size then pick a chunk size
    // ceiling based on the worst case. Each test takes time so there's trade-off.
    // This attempts to prevent small sites with very large pages from OOMing.
    // This heuristic could still fail if it randomly grabs the smallest nodes.
    // TODO: test a few nodes per each type instead of from all nodes
    const nodesToTest = 11; // Very arbitrary number
    const valueCount = values.length;
    const step = Math.max(1, Math.ceil(valueCount / nodesToTest));
    let maxSize = 0;
    for (let i = 0; i < valueCount; i += step) {
        const size = v8_1.default.serialize(values[i]).length;
        maxSize = Math.max(size, maxSize);
    }
    // Sends a warning once if any of the chunkSizes exceeds approx 500kb limit
    if (showMaxSizeWarning && maxSize > 500000) {
        reporter_1.default.warn(`The size of at least one page context chunk exceeded 500kb, which could lead to degraded performance. Consider putting less data in the page context.`);
    }
    // Max size of a Buffer is 2gb (yeah, we're assuming 64bit system)
    // https://stackoverflow.com/questions/8974375/whats-the-maximum-size-of-a-node-js-buffer
    // Use 1.5gb as the target ceiling, allowing for some margin of error
    return Math.floor((1.5 * 1024 * 1024 * 1024) / maxSize);
}
function prepareCacheFolder(targetDir, contents) {
    // Temporarily save the nodes and pages and remove them from the main redux store
    // This prevents an OOM when the page nodes collectively contain too much data
    const nodesMap = contents.nodes;
    contents.nodes = undefined;
    contents.schemaCustomization = (0, flatted_1.stringify)(contents.schemaCustomization);
    (0, fs_extra_1.writeFileSync)(reduxSharedFile(targetDir), v8_1.default.serialize(contents));
    // Now restore them on the redux store
    contents.nodes = nodesMap;
    if (nodesMap) {
        if (nodesMap.size === 0) {
            // Nodes are actually stored in LMDB.
            // But we need at least one node in redux state to workaround the warning above:
            // "Cache exists but contains no nodes..." (when loading cache).
            const dummyNode = {
                id: `dummy-node-id`,
                parent: ``,
                children: [],
                internal: {
                    type: `DummyNode`,
                    contentDigest: `dummy-node`,
                    counter: 0,
                    owner: ``,
                },
                fields: [],
            };
            nodesMap.set(dummyNode.id, dummyNode);
        }
        // Now store the nodes separately, chunk size determined by a heuristic
        const values = [...nodesMap.entries()];
        const chunkSize = guessSafeChunkSize(values);
        const chunks = Math.ceil(values.length / chunkSize);
        for (let i = 0; i < chunks; ++i) {
            (0, fs_extra_1.writeFileSync)(reduxChunkedNodesFilePrefix(targetDir) + i, v8_1.default.serialize(values.slice(i * chunkSize, i * chunkSize + chunkSize)));
        }
    }
}
function safelyRenameToBak(cacheFolder) {
    // Basically try to work around the potential of previous renamed caches
    // not being removed for whatever reason. _That_ should not be a blocker.
    const tmpSuffix = `.bak`;
    let suffixCounter = 0;
    let bakName = cacheFolder + tmpSuffix; // Start without number
    while ((0, fs_extra_1.existsSync)(bakName)) {
        ++suffixCounter;
        bakName = cacheFolder + tmpSuffix + suffixCounter;
    }
    (0, fs_extra_1.moveSync)(cacheFolder, bakName);
    return bakName;
}
function writeToCache(contents) {
    // Note: this should be a transactional operation. So work in a tmp dir and
    // make sure the cache cannot be left in a corruptable state due to errors.
    const tmpDir = (0, fs_extra_1.mkdtempSync)(path_1.default.join(os_1.default.tmpdir(), `reduxcache`)); // linux / windows
    prepareCacheFolder(tmpDir, contents);
    // Replace old cache folder with new. If the first rename fails, the cache
    // is just stale. If the second rename fails, the cache is empty. In either
    // case the cache is not left in a corrupt state.
    const reduxCacheFolder = getReduxCacheFolder();
    let bakName = ``;
    if ((0, fs_extra_1.existsSync)(reduxCacheFolder)) {
        // Don't drop until after swapping over (renaming is less likely to fail)
        bakName = safelyRenameToBak(reduxCacheFolder);
    }
    // The redux cache folder should now not exist so we can rename our tmp to it
    (0, fs_extra_1.moveSync)(tmpDir, reduxCacheFolder);
    // Now try to yolorimraf the old cache folder
    try {
        if (bakName !== ``) {
            (0, fs_extra_1.removeSync)(bakName);
        }
    }
    catch (e) {
        reporter_1.default.warn(`Non-fatal: Deleting the old cache folder failed, left behind in \`${bakName}\`. Rimraf reported this error: ${e}`);
    }
}
